# Advanced MLP - 2
- Putting it altogether

In [11]:
import matplotlib.pyplot as plt
import numpy as np

from sklearn.ensemble import VotingClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from keras.datasets import mnist
from keras.wrappers.scikit_learn import KerasClassifier
from keras.datasets import mnist
from keras.models import Sequential
from keras.utils.np_utils import to_categorical
from keras.models import Sequential
from keras.layers import Activation, Dense, BatchNormalization, Dropout
from keras import optimizers

## Load Dataset

In [12]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [13]:
# reshaping X data: (n, 28, 28) => (n, 784)
X_train = X_train.reshape((X_train.shape[0], X_train.shape[1] * X_train.shape[2]))
X_test = X_test.reshape((X_test.shape[0], X_test.shape[1] * X_test.shape[2]))

In [22]:
# We use all training data and validate on all test data
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(60000, 784)
(10000, 784)
(60000,)
(10000,)


## Training & Validating Model
- Measures to improve training is applied simultaneously
    - More training set
    - Weight Initialization scheme
    - Nonlinearity (Activation function)
    - Optimizers: adaptvie
    - Batch Normalization
    - Dropout (Regularization)
    - Model Ensemble

In [16]:
def mlp_model():
    model = Sequential()
    
    model.add(Dense(50, input_shape = (784, ), kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))    
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(10, kernel_initializer='he_normal'))
    model.add(Activation('softmax'))
    
    adam = optimizers.Adam(lr = 0.001)
    model.compile(optimizer = adam, loss = 'categorical_crossentropy', metrics = ['accuracy'])
    
    return model

In [17]:
# create 5 models to ensemble
model1 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model2 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model3 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model4 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model5 = KerasClassifier(build_fn = mlp_model, epochs = 100)

In [18]:
ensemble_clf = VotingClassifier(estimators = [('model1', model1), ('model2', model2), ('model3', model3), ('model4', model4), ('model5', model5)], voting = 'soft')

In [19]:
ensemble_clf.fit(X_train, y_train)

Epoch 1/100
60000/60000 [==============================] - 18s - loss: 0.7110 - acc: 0.7858    
Epoch 2/100
60000/60000 [==============================] - 15s - loss: 0.4120 - acc: 0.8827    
Epoch 3/100
60000/60000 [==============================] - 14s - loss: 0.3546 - acc: 0.8995    
Epoch 4/100
60000/60000 [==============================] - 15s - loss: 0.3187 - acc: 0.9095    
Epoch 5/100
60000/60000 [==============================] - 14s - loss: 0.2930 - acc: 0.9167    
Epoch 6/100
60000/60000 [==============================] - 14s - loss: 0.2768 - acc: 0.9214    
Epoch 7/100
60000/60000 [==============================] - 15s - loss: 0.2605 - acc: 0.9262    
Epoch 8/100
60000/60000 [==============================] - 15s - loss: 0.2510 - acc: 0.9284    
Epoch 9/100
60000/60000 [==============================] - 15s - loss: 0.2450 - acc: 0.9306    
Epoch 10/100
60000/60000 [==============================] - 15s - loss: 0.2352 - acc: 0.9327    
Epoch 11/100
60000/60000 [=============

60000/60000 [==============================] - 16s - loss: 0.1514 - acc: 0.9558    
Epoch 70/100
60000/60000 [==============================] - 14s - loss: 0.1495 - acc: 0.9569    
Epoch 71/100
60000/60000 [==============================] - 16s - loss: 0.1488 - acc: 0.9565    
Epoch 72/100
60000/60000 [==============================] - 17s - loss: 0.1493 - acc: 0.9564    
Epoch 73/100
60000/60000 [==============================] - 15s - loss: 0.1478 - acc: 0.9564    
Epoch 74/100
60000/60000 [==============================] - 15s - loss: 0.1514 - acc: 0.9557    
Epoch 75/100
60000/60000 [==============================] - 15s - loss: 0.1514 - acc: 0.9561    
Epoch 76/100
60000/60000 [==============================] - 15s - loss: 0.1507 - acc: 0.9554    
Epoch 77/100
60000/60000 [==============================] - 18s - loss: 0.1489 - acc: 0.9562    
Epoch 78/100
60000/60000 [==============================] - 19s - loss: 0.1511 - acc: 0.9558    
Epoch 79/100
60000/60000 [=================

60000/60000 [==============================] - 18s - loss: 0.1732 - acc: 0.9488    
Epoch 38/100
60000/60000 [==============================] - 18s - loss: 0.1702 - acc: 0.9515    
Epoch 39/100
60000/60000 [==============================] - 17s - loss: 0.1686 - acc: 0.9519    
Epoch 40/100
60000/60000 [==============================] - 18s - loss: 0.1698 - acc: 0.9512    
Epoch 41/100
60000/60000 [==============================] - 18s - loss: 0.1653 - acc: 0.9516    
Epoch 42/100
60000/60000 [==============================] - 18s - loss: 0.1644 - acc: 0.9517    
Epoch 43/100
60000/60000 [==============================] - 18s - loss: 0.1671 - acc: 0.9526    
Epoch 44/100
60000/60000 [==============================] - 18s - loss: 0.1640 - acc: 0.9522    
Epoch 45/100
60000/60000 [==============================] - 18s - loss: 0.1638 - acc: 0.9533    
Epoch 46/100
60000/60000 [==============================] - 18s - loss: 0.1648 - acc: 0.9526    
Epoch 47/100
60000/60000 [=================

VotingClassifier(estimators=[('model1', <keras.wrappers.scikit_learn.KerasClassifier object at 0x00000240E29D87F0>), ('model2', <keras.wrappers.scikit_learn.KerasClassifier object at 0x00000240E29D8828>), ('model3', <keras.wrappers.scikit_learn.KerasClassifier object at 0x00000240E29D8898>), ('model4', <keras.wrappers.scikit_learn.KerasClassifier object at 0x00000240E29D88D0>), ('model5', <keras.wrappers.scikit_learn.KerasClassifier object at 0x00000240E29D8908>)],
         n_jobs=1, voting='soft', weights=None)

In [20]:
y_pred = ensemble_clf.predict(X_test)

 9920/10000 [============================>.] - ETA: 0s

In [24]:
print('Acc: ', accuracy_score(y_pred, y_test))

Acc:  0.9801


Final model produces over **98% of accuracy**, which is rather improved from before